### ISSUE TITLE
No audio with AAC tracks on old Android devices

### ISSUE THREAD

**Issue by Paco8** [peak3d/inputstream.adaptive#1291]
### Describe the bug
Here is a clear and concise description of what the problem is:

I installed Kodi in an old Android Smartphone with Android 6.0 and I noticed that some addons don't have sound. For example this happens with the hbo max addon. I think the problem happens with encrypted AAC audio tracks (DD+ have sound). If I play a video with no DRM with an AAC track, there's audio. I could reproduce the problem in another Smartphone with Android 8.1. However a Smartphone with Android 11 doesn't have the problem.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

There should be audio.

## Actual Behavior

There's no audio and the log is full with this error:

```
error <general>: CDVDAudioCodecAndroidMediaCodec::GetData ExceptionCheck; dequeueOutputBuffer
                                                    java.lang.IllegalStateException
```

### To Reproduce

1. You need an old Android device.
2. Play this strm file:

```
#KODIPROP:inputstream=inputstream.adaptive
#KODIPROP:inputstream.adaptive.manifest_type=ism
#KODIPROP:inputstream.adaptive.license_type=com.widevine.alpha
#KODIPROP:inputstream.adaptive.license_key=https://ampvideos.keydelivery.mediaservices.windows.net/Widevine/?kid=096d74b9-362d-4110-949b-ec0f5694f750||R{SSM}|
https://ampvideos.streaming.mediaservices.windows.net/9b77ed82-94af-41bc-b08b-537e7a530eb6/sintel_trailer-1080p_fp.ism/manifest
```

### Debuglog

The debuglog can be found here:

https://paste.kodi.tv/opeyutebok.kodi

### MPD/M3U8s/ISM

An example or copy of a manifest (or manifests for HLS - master and variants) can be found here:

https://ampvideos.streaming.mediaservices.windows.net/9b77ed82-94af-41bc-b08b-537e7a530eb6/sintel_trailer-1080p_fp.ism/manifest

### Screenshots
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

I tested with Kodi 21 alpha with the same result.

### Your Environment
Used Operating system:

 - [x] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Android 6.0.0
 - Kodi version: 20.1
 - inputstream.adaptive version: 20.3.9

*note: Once the issue is made we require you to update it with new information should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by CastagnaIT** [comment]
look like that an android exception related you should provide also the android logcat

**Comment by Paco8** [comment]
[logcat.zip](https://github.com/xbmc/inputstream.adaptive/files/11770579/logcat.zip)

**Comment by CastagnaIT** [comment]
looks like a more lower level problem not related to ISA
e.g.
```
06-16 14:58:18.813   283  9178 D MtkOmxRawDec: Init RAW PCM Decoder !!!
06-16 14:58:18.813   283  9178 E MtkOmxRawDec: MtkOmxAacDec::DecodeAudio RawPCM_INVALID_BITSTREAM!
06-16 14:58:18.813   283  9178 D MtkOmxRawDec: --- OMX_EventPortSettingsChanged ---
06-16 14:58:18.813 29128  9176 E ACodec  : [OMX.MTK.AUDIO.DECODER.RAW] ERROR(0x80001005)
06-16 14:58:18.814 29128  9176 E ACodec  : signalError(omxError 0x80001005, internalError -2147483648)
06-16 14:58:18.814 29128  9176 W ACodec  : onEvent--Audio OMX Bad Parameter!!
```
i have already see something similar on other older android devices (maybe was android 7, idr) with similar driver, and no solutions found then here i stop

@joseluismarti when you will have time can you try to see the audio errors on log?
if its something that you are able to fix in kodi or not

**Comment by joseluismarti** [comment]
I've tested the stream you propose on Android 7.1.1 phone and it looks and sounds perfect.

This error is a low level failure in the codec provided by the manufacturer of this device. It does not manage correctly some parameter, something is missing or has some internal error, I do not think we can do anything.

Check that you have the latest version provided by the manufacturer. Another option would be to try an alternative ROM in case it includes more up to date codecs.

**Comment by Paco8** [comment]
Unfortunately, there's no update available for this device. However, there's something I don't understand: if it's a bug in the firmware, why does the problem seem to occur only with Kodi?

I developed this addon [plugin.video.orange.spain](https://github.com/Paco8/plugin.video.orange.spain), and it has the issue of no audio, but the official Orange TV app does have audio. I analyzed the traffic from the app, and it uses the same manifest. Somehow the official app manages to play the audio.

**Comment by CastagnaIT** [comment]
i think its possible that the official app use a different audio decoder than kodi
if i have not read wrong kodi use `OMX.MTK.AUDIO.DECODER.RAW`
and its possible that the official app use `OMX.google.raw.decoder`

i have also found this discussion https://github.com/google/ExoPlayer/issues/2873 where another device have same errors
and the solution is prefer `OMX.google.raw.decoder` over `OMX.MTK.AUDIO.DECODER.RAW` for old android versions

@joseluismarti this could have sense? do you think we can try implement/test a similar workaround also for kodi? or add a fallback way?

**Comment by joseluismarti** [comment]
Yes it might make sense, although it would have to be tested to be sure. The device should have another compatible codec available.

A blacklist could be added so that certain codecs could be disabled on demand.

**Comment by joseluismarti** [comment]
@Paco8 I've prepared a [test apk](https://drive.google.com/file/d/14LXjrGCZ1CNU1RLoZr1Q9q7Z4lzZAvf9/view?usp=sharing), if both codecs are available on the device, then use the one from google.

First you have to uninstall Kodi because the signature of the app is different. Be careful, if necessary make a backup since uninstalling deletes everything.

**Comment by Paco8** [comment]
Yes, now there's audio in that device. Thank you.

**Comment by joseluismarti** [comment]
Can you check if it also fixes it on the device with Android 8.1?

**Comment by CastagnaIT** [comment]
i found the old issue so this potentially can fix also audio problems on #775

**Comment by Paco8** [comment]
> Can you check if it also fixes it on the device with Android 8.1?

Yes, it works as well.
